### PR TITLE
Get rid of std::bind.

### PR DESCRIPTION
**Pull request by wohe**
`std::bind` is bug prone and should be avoided.
Lambdas are a more general and safer replacement.

### CODE REVIEW

**Comment by wohe** [comment]
E.g. https://www.youtube.com/watch?v=zt7ThwVfap0&t=1940s
https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines

File: cartographer/mapping/internal/constraints/constraint_builder_2d_test.cc [hunk 1]
```diff
@@ -60,7 +60,9 @@ TEST_F(ConstraintBuilder2DTest, CallsBack) {
   EXPECT_CALL(mock_, Run(::testing::IsEmpty()));
   constraint_builder_->NotifyEndOfNode();
   constraint_builder_->WhenDone(
-      std::bind(&MockCallback::Run, &mock_, std::placeholders::_1));
+      [this](const constraints::ConstraintBuilder2D::Result& result) {
```
Comment (gaschler) [inline cartographer/mapping/internal/constraints/constraint_builder_2d_test.cc]: please check if capturing &mock_ is enough, also below

Comment (gaschler) [reply cartographer/mapping/internal/constraints/constraint_builder_2d_test.cc]: ok, then no change, less noise is better

**Comment by wohe** [comment]
@wally-the-cartographer merge

**Comment by wally-the-cartographer** [comment]
Merge requested by authorized user wohe. Merge queue now has a length of 2.

**Pull request merged**